### PR TITLE
INT-3566: JMS Recover Connection During Req/Resp

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-3566

When using a named replyTo queue/topic if the connection was lost between the
request and reply, the connection was not recovered.

Retry consuming until the timeout expires.

### CODE REVIEW

File: spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java [hunk 1]
```diff
@@ -864,18 +861,89 @@ else if (replyTo instanceof TemporaryQueue || replyTo instanceof TemporaryTopic)
...
+		long now = System.currentTimeMillis();
+		JMSException exception = null;
+		boolean isTemporaryReplyTo = replyTo instanceof TemporaryQueue || replyTo instanceof TemporaryTopic;
+		try {
+			do {
+				try {
+					messageConsumer = consumerSession.createConsumer(replyTo, messageSelector);
+					javax.jms.Message reply = receiveReplyMessage(messageConsumer);
+					if (reply == null) {
+						if (this.receiveTimeout < 0 || now + this.receiveTimeout > System.currentTimeMillis()) {
```
Comment (artembilan) [inline spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java]: Minor, of course, but it isn't `now` here already after several iterations.
How about `receiveStart` ?

File: spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java [hunk 2]
```diff
@@ -864,18 +861,89 @@ else if (replyTo instanceof TemporaryQueue || replyTo instanceof TemporaryTopic)
...
+								return null;
+							}
+						}
+					}
+					while (!isTemporaryReplyTo &&
+							(this.receiveTimeout < 0 || now + this.receiveTimeout > System.currentTimeMillis()));
+				}
+			}
+			while (!isTemporaryReplyTo &&
+					(this.receiveTimeout < 0 || now + this.receiveTimeout > System.currentTimeMillis()));
```
Comment (artembilan) [inline spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java]: Looks like we should extract a variable in the beginning for: `this.receiveTimeout < 0 || now + this.receiveTimeout`

File: spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java [hunk 3]
```diff
@@ -864,18 +861,89 @@ else if (replyTo instanceof TemporaryQueue || replyTo instanceof TemporaryTopic)
...
+				throw exception;
+			}
+		}
+		finally {
+			if (consumerSession != session) {
+				JmsUtils.closeSession(consumerSession);
+				JmsUtils.closeConnection(consumerConnection);
+			}
+			else {
+				JmsUtils.closeMessageConsumer(messageConsumer);
```
Comment (artembilan) [inline spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java]: Why only in the `else`?
Does `JmsUtils.closeConnection(consumerConnection)` close `consumer` as well ?

File: spring-integration-jms/src/test/java/org/springframework/integration/jms/JmsOutboundGatewayTests.java [hunk 4]
```diff
@@ -154,4 +181,104 @@ public Message answer(InvocationOnMock invocation) throws Throwable {
 			gateway.stop();
 		}
 	}
+
+	@Test
+	public void testConnectionBreakOnReplyMessageIdCorrelation() throws Exception {
+		CachingConnectionFactory connectionFactory1 = new CachingConnectionFactory(
+				new ActiveMQConnectionFactory("tcp://10.0.0.3:61616"));
```
Comment (artembilan) [inline spring-integration-jms/src/test/java/org/springframework/integration/jms/JmsOutboundGatewayTests.java]: @garyrussell , I don't know how have you tested it. (Looks like you have some external ActiveMQ Broker).
Anyway these tests (with external URL) don't pass even for me, not only for Travis :smile: 
From other side this `61616` fixed port is dangerous.
This one works for me: https://github.com/artembilan/spring-integration/commit/f2840a92a940084dde7650e03829185ad2da0317

Comment (garyrussell) [inline spring-integration-jms/src/test/java/org/springframework/integration/jms/JmsOutboundGatewayTests.java]: Looks like I missed adding my local change to change it back; re-pushed - I'll look at your other suggestions later.

**Comment by garyrussell** [comment]
Pushed

**Comment by artembilan** [comment]
LGTM, if that
```
org.springframework.integration.jms.JmsOutboundGatewayTests > classMethod FAILED
    javax.management.InstanceAlreadyExistsException at JmsOutboundGatewayTests.java:83
```
will be easy to fix on merge.

**Comment by garyrussell** [comment]
? There's no JMX in the JMS module

**Comment by artembilan** [comment]
Right, I don't mind, but Travis is barking somehow...

**Comment by artembilan** [comment]
Merging, by the way...

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/585fb0349c461f7d88e838b8370492b971b088dd

**Pull request closed without merging**